junebug/logging_service.py

```diff
@@ -122,3 +123,61 @@ def stopService(self):
 
     def registered(self):
         return self.log_observer in self.logger.observers
+
+
+def reverse_read(filename, buf):
+    '''
+    Read the non-blank lines from a file in reverse order.
+
+    :param str filename: The path of the file to be read.
+
+    :returns:
+        A generator that yields each line from the file in reverse order as a
+        string.
+    '''
+    with open(filename) as f:
+        f.seek(0, os.SEEK_END)
+        total_size = remaining_size = f.tell()
+        offset = 0
+        incomplete_line = None
+        while remaining_size > 0:
+            offset = min(total_size, offset + buf)
+            f.seek(-offset, os.SEEK_END)
+            data = f.read(min(remaining_size, buf))
+            remaining_size -= buf
+            lines = data.split('\n')
+            if incomplete_line:
+                lines[-1] += incomplete_line
+            incomplete_line = lines.pop(0)
+            for l in lines[::-1]:
+                if l is not '':
+                    yield l
+        if incomplete_line:
+            yield incomplete_line
+
+
+def read_logs(logfile, lines, buf=4096):
+    '''
+    Reads log files to fetch the last N logs.
+
+    :param logfile: The LogFile to read from.
+    :type logfile: :class:`twisted.python.logfile.LogFile`
+    :param int lines: Maximum number of lines to read.
+    :param int buffer: The size of the read buffer.
+
+    :returns: list -- Each item is a dictionary representing a log entry.
+    '''
+    logs = []
+
+    # Read the current log file
+    for line in reverse_read(logfile.path, buf):
+        logs.append(json.loads(line))
+        if len(logs) == lines:
+            return logs
+    # Read the rotated log files
+    for file_num in logfile.listLogs():
```

Comment:
Or a small helper generator function that yields the filenames if that
seems clearer?